src/ray/core_worker/core_worker.cc

```diff
@@ -239,6 +246,37 @@ void CoreWorkerProcess::RunTaskExecutionLoop() {
   instance_.reset();
 }
 
+void CoreWorkerProcess::RunStatsService() {
+  absl::MutexLock lock(&CoreWorkerProcess::stats_initialization_mutex_);
+  enable_stats_count_++;
```

Comment:
What is this value for? Since stats::Init is already thread-safe and initialize only once, we don't need any of them technically?

Reply:
stats::init is thread-safe, but we can't shutdown stats after the last core worker deconstructed. So we save the number of core worker, then shutdown when last one exits.

Reply:
But, isn't `CoreWorkerProcess::Shutdown` called only once when the process exits?

Reply:
For example, for Python workers, the Cython destructor is calling this method (right before shutdown).

Reply:
Sorry if I am missing something from Java worker cases

Reply:
@rkooo567 According to description of @kfstorm and core worker's future plan, java worker and diriver could call multiple  `CoreWorker::Constructor` and `CoreWorker::Shutdown`.  Driver-wise, it will be initialized and shutdown if ray.init and ray.shutdown are invoked.

Reply:
For worker-wise, CoreWorkerProcess::Shutdown never be called initiatively, which makes no shutdown action for stats module, then crashed and failed after unit test main thread exits.

Reply:
`CoreWorkerProcess::Initialize` can only be called once by workers, but not drivers. A driver may initialize and shutdown Ray multiple times. A typical use case is Python and Java tests. And `CoreWorkerProcess::Shutdown` is meant to be called by driver only. (the documentation is at the top of `class CoreWorkerProcess` declaration.) A worker can shut down the `CoreWorkerProcess` by returning a system exit status in the task execution callback (also documented), and in this way, the `CoreWorkerProcess::RunTaskExecutionLoop` will finish and the worker process is ready to quit.

A good way to avoid this confusion is to put the `ray::Stats::Init` and `ray::Stats::Shutdown` to the constructor and destructor of `CoreWorkerProcess`.